core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -222,4 +230,89 @@ public ByteSizeValue paramAsSize(String key, ByteSizeValue defaultValue) {
+     * {@linkplain RestHandler} treats the {@code source} parameter like the body content.
+     */
+    public final boolean hasContentOrSourceParam() {
+        return hasContent() || hasParam("source");
+    }
+
+    /**
+     * The {@linkplain XContentType} of either the request body or the {@code source} param if either exists and can be guessed, otherwise
+     * {@code null}. Only use this if you need to do something different if the request is plain text. Use
+     * {@link #contentOrSourceParamParser()} if you need a parser.
+     */
+    @Nullable
+    public final XContentType contentOrSourceParamXContentType() {
+        BytesReference content = contentOrSource();
+        return content == null ? null : XContentFactory.xContentType(content);
+    }
+
+    /**
+     * A parser for the contents of this request if it has contents, otherwise a parser for the {@code source} parameter if there is one,
+     * otherwise throws an {@link ElasticsearchParseException}. Use {@link #contentOrSourceParamParserOrNull()} instead if you need to
+     * handle the absence of a body gracefully.
+     */
+    public final XContentParser contentOrSourceParamParser() throws IOException {
+        BytesReference content = contentOrSource();
+        if (content.length() == 0) {
+            throw new ElasticsearchParseException("Body required");
+        }
+        return XContentFactory.xContent(content).createParser(content);
+    }
+
+    /**
+     * A parser for the contents of this request if it has contents, otherwise a parser for the {@code source} parameter if there is one,
+     * otherwise {@code null}. Use {@link #contentOrSourceParamParser()} if you should throw an exception back to the user when there isn't
+     * a body. See also {@link #withContentOrSourceParamParserOrNull(IOConsumer)} which may cleaner to use.
+     */
+    @Nullable
+    public final XContentParser contentOrSourceParamParserOrNull() throws IOException {
+        BytesReference content = contentOrSource();
+        if (content.length() == 0) {
+            return null;
+        }
+        return XContentFactory.xContent(content).createParser(content);
+    }
+
+    /**
+     * Helper around {@link #contentOrSourceParamParserOrNull()} that closes the parser for you. This is just syntactic sugar to ease
+     * handling the nullable parser.
+     */
+    @Nullable
+    public final void withContentOrSourceParamParserOrNull(IOConsumer<XContentParser> withParser) throws IOException {
+        XContentParser parser = contentOrSourceParamParserOrNull();
+        try {
+            withParser.accept(parser);
+        } finally {
+            IOUtils.close(parser);
+        }
+    }
+
+    /**
+     * Get the content of the request or the contents of the {@code source} param. Prefer {@link #contentOrSourceParamParser()} if possible
+     * to centralize parser construction.
+     */
+    public final BytesReference contentOrSourceParam() {
```

Comment:
Originally I'd exposed this as `@Deprecated`, that is why the two methods. I feel like it *should* be deprecated because we want to work with parsers where we can and we're parsing on the coordinating node, but too many APIs use it.